design/worker-pods.md

```diff
@@ -0,0 +1,161 @@
+
+The spec for the worker pod will look like the following:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    component: velero
+    velero.io/backup-name: <BACKUP_NAME>
+  name: backup-<BACKUP_NAME>
+  namespace: velero
+  ownerReferences:
+  - apiVersion: velero.io/v1
+    controller: true
+    kind: Backup
+    name: <BACKUP_NAME>
+    uid: <BACKUP_UID>
+spec:
+  containers:
+  - args:
+    - <BACKUP_NAME>
+    - --client-burst=<VAL>
+    - --client-qps=<VAL>
+    - --default-backup-storage-location=<VAL>
+    - --default-backup-ttl=<VAL>
+    - --default-volume-snapshot-locations=<VAL>
+    - --log-level=<VAL>
+    - --restic-timeout=<VAL>
+    command:
+    - /velero
+    - backup
+    - run
+    env:
+    - name: VELERO_SCRATCH_DIR
+      value: /scratch
+    - name: AWS_SHARED_CREDENTIALS_FILE
+      value: /credentials/cloud
+    image: <VELERO_SERVER_IMAGE>
+    imagePullPolicy: IfNotPresent
+    name: velero
+    volumeMounts:
+    - mountPath: /plugins
+      name: plugins
+    - mountPath: /scratch
+      name: scratch
+    - mountPath: /credentials
+      name: cloud-credentials
+    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
+      name: velero-token-plwsw
+      readOnly: true
+  serviceAccount: velero
+  volumes:
+  - emptyDir: {}
+    name: plugins
+  - emptyDir: {}
+    name: scratch
+  - name: cloud-credentials
+    secret:
+      defaultMode: 420
+      secretName: cloud-credentials
+  - name: velero-token-plwsw
+    secret:
+      defaultMode: 420
+      secretName: velero-token-plwsw
+```
+
+### Updates to `velero backup logs`
+
+The `velero backup logs` command will be modified so that if a backup is in progress, the logs are gotten from the worker pod's stdout (essentially proxying to `kubectl -n velero logs POD_NAME`). A  `-f/--follow` flag will be added to allow streaming of logs for an in-progress backup.
+
+Once a backup is complete, the logs will continue to be uploaded to object storage, and `velero backup logs` will still fetch them from there.
+
+### Changes to backup deletion controller
+
+The Velero server keeps a record of current in-progress backups and disallows them from being deleted. This will need to be updated to account for backups running in worker pods. The most sensible approach is for the backup deletion controller to first check the backup's status via the API, and to allow deletion to proceed if it's not new/in-progress. If the backup is new or in-progress, the controller can check to see if a worker pod is running for the backup. If so, the deletion is disallowed because the backup is actively being processed. If there is no worker pod for the backup, the backup can be deleted as it's not actively being processed.
+
+### Open items
+
+- It probably makes sense to use Kubernetes Jobs to control worker pods, rather than directly creating "bare pods". The Job will ensure that a worker pod successfully runs to completion.
```

Comment:
One of the benefits of Jobs over Pods is being able to configure the maximum number of retries, but if we don't need to retry failed backups then a Pod should suffice. Do we retry backups today?

Reply:
As of today, we just fail backups and restores. Retries may be helpful; I'd appreciate community feedback on that.

Reply:
In my mind there are two different scenarios to consider here:

1. The backup completes, with a final phase of "Failed" or "PartiallyFailed". I would not currently consider this grounds for a retry.
2. The worker pod gets evicted or otherwise forcibly terminated before the backup has had a chance to complete. In this case I *would* want to retry.